otherlibs/str/strstubs.c

```diff
@@ -89,10 +94,81 @@ struct re_group {
 /* Record positions reached during matching; used to check progress
    in repeated matching of a regexp. */
 #define NUM_REGISTERS 64
-static unsigned char * re_register[NUM_REGISTERS];
+static pthread_key_t re_register_key;
+
+/* The initial backtracking stack key */
+static pthread_key_t initial_stack_key;
+
+static void free_backtrack_stack(struct backtrack_stack *);
+
+static struct backtrack_stack *get_backtrack_stack()
+{
+  struct backtrack_stack *stack = pthread_getspecific(initial_stack_key);
+
+  if (stack == NULL) {
+    stack = caml_stat_alloc(sizeof(struct backtrack_stack));
+    stack->previous = NULL;
+    pthread_setspecific(initial_stack_key, stack);
+    return stack;
+  }
+  else
+    return stack;
+}
+
+static unsigned char **get_re_register()
+{
+  unsigned char **re_register = pthread_getspecific(re_register_key);
 
-/* The initial backtracking stack */
-static struct backtrack_stack initial_stack = { NULL, };
+  if (re_register == NULL) {
+    re_register =  caml_stat_alloc(sizeof(unsigned char *) * NUM_REGISTERS);
```

Comment:
Should we add error handling if the allocation fails?

Reply:
it's been pointed out that the error handling is at least how it is for the rest of the file. 
we can leave it and merge this PR as is.